hpcbench/benchmark/minigemm.py

```diff
@@ -0,0 +1,261 @@
+            << tdiff.count() << "s\\n";
+  std::cout << gflop/tdiff.count() << " GFLOP/s\\n";
+
+#ifdef DBG_PRINT
+  for(int i = 0; i < min(m, 5); ++i) {
+    for(int j = 0; j < min(n, 5); ++j) {
+      std::cout << Cs[0][j + i*n] << "  ";
+    }
+    std::cout << std::endl;
+  }
+#endif
+
+  double chk;
+  double sgn = 1.0;
+  for(int j = 0; j < nthreads; j++) {
+    for(int i = 0; i < m*n; ++i) {
+      sgn *= -1.0;
+      chk += sgn*Cs[j][i];
+    }
+  }
+  std::cout << "Check value: " << chk << std::endl;
+
+  for(int j = 0; j < 2; j++) {
+    mkl_free(As[j]);
+    mkl_free(Bs[j]);
+    mkl_free(Cs[j]);
+  }
+  free(As);
+  free(Bs);
+  free(Cs);
+
+  return 0;
+}
+"""
+
+COMPILE_SCRIPT = """#!/bin/bash -l
+make DEFINES="{opts}"
+"""
+
+MAKEFILE = """CXX=icpc
+DEFINES=
+CXXFLAGS=-qopt-streaming-stores always -O3 -mkl -std=c++11 -qopenmp
+LDFLAGS=
+
+CPPFILES=minigemm.cpp
+OBJ=$(CPPFILES:.cpp=.o)
+
+all: minigemm
+
+.cpp.o: $<
+	$(CXX) -c $< $(DEFINES) $(CXXFLAGS)
+
+minigemm: $(OBJ)
+	$(CXX) -o $@ $^ $(CXXFLAGS) $(LDFLAGS)
+
+clean:
+	$(RM) minigemm *.o
+
+.PHONY: clean
+
+"""
+
+
+class MiniGEMM(Benchmark):
+    """A mini benchmark performing MKL based generalized matrix multiplies
+    to determine the node's peak floating point performance.
+    """
+    name = 'minigemm'
+    description = "DGEMM mini benchmark"
+    COMPILE_PARAMS = []
+    DEFAULT_OPENMP_THREADS = 1
+    DEFAULT_NESTED_MKL_THREADS = 4
+
+    def __init__(self):
+        super(MiniGEMM, self).__init__(
+              attributes=dict(compile=self.COMPILE_PARAMS,
+                              omp_threads=self.DEFAULT_OPENMP_THREADS,
+                              mkl_threads=self.DEFAULT_NESTED_MKL_THREADS
+        ))
+
+    @cached_property
+    def compile(self):
+        return self.attributes['compile']
+
+    @cached_property
+    def omp_threads(self):
+        return str(self.attributes['omp_threads'])
+
+    @cached_property
+    def mkl_threads(self):
+        return str(self.attributes['mkl_threads'])
+
+    def execution_matrix(self, context):
+        del context  # unused
+        yield dict(
+            category=self.name,
+            command=['./minigemm'],
+            environment=dict(
+                MKL_DYNAMIC='false',
+                OMP_NESTED='true',
+                OMP_PROC_BIND='spread,close',
+                OMP_NUM_THREADS=self.omp_threads,
+                MKL_NUM_THREADS=self.mkl_threads,
+            ),
+        )
+
+    def _compile(self, execution):
+        with open('minigemm.cpp', 'w') as ostr:
+            print(SOURCE, file=ostr)
+        with open('Makefile', 'w') as ostr:
+            print(MAKEFILE, file=ostr)
+        opt_str = ' '.join(self.compile)
+        with open('compile.sh', 'w') as ostr:
+            print(COMPILE_SCRIPT.format(opts=opt_str), file=ostr)
+        st = os.stat('compile.sh')
+        os.chmod('compile.sh', st.st_mode | stat.S_IEXEC)
+        ret = subprocess.call(['./compile.sh'])
```

Comment:
Maybe use `subprocess.check_call` instead of `subprocess.call` (to make sure to have an unreadable stack trace ;)